pyleco/core/serialization.py

```diff
@@ -0,0 +1,99 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import json
+from typing import Optional, Tuple
+
+from uuid_extensions import uuid7  # as long as uuid does not yet support UUIDv7
+from pydantic import BaseModel
+
+
+def create_header_frame(conversation_id: Optional[bytes] = None,
+                        message_id: Optional[bytes] = None,
+                        message_type: Optional[bytes] = None) -> bytes:
+    """Create the header frame.
+
+    :param bytes conversation_id: ID of the conversation.
+    :param bytes message_id: Message ID of this message, must not contain.
+    :return: header frame.
+    """
+    if conversation_id is None:
+        conversation_id = generate_conversation_id()
+    elif len(conversation_id) != 16:
+        raise ValueError("Length of 'conversation_id' is not 16 bytes.")
+    if message_id is None:
+        message_id = b"\x00" * 3
+    elif len(message_id) != 3:
+        raise ValueError("Length of 'message_id' is not 3 bytes.")
+    if message_type is None:
+        message_type = b"\x00"
+    elif len(message_type) != 1:
+        raise ValueError("Length of 'message_type' is not 1 bytes.")
+    return b"".join((conversation_id, message_id, message_type))
+
+
+def split_name(name: bytes, node: bytes = b"") -> Tuple[bytes, bytes]:
+    """Split a sender/receiver name with given default node."""
+    s = name.split(b".")
+    n = s.pop(-1)
+    return (s.pop() if s else node), n
+
+
+def split_name_str(name: str, node: str = "") -> Tuple[str, str]:
+    """Split a sender/receiver name with given default node."""
+    s = name.split(".")
+    n = s.pop(-1)
+    return (s.pop() if s else node), n
+
+
+def interpret_header(header: bytes) -> Tuple[bytes, bytes, bytes]:
+    """Interpret the header frame.
+
+    :return: conversation_id, message_id, message_type
+    """
+    conversation_id = header[:16]
+    message_id = header[16:19]
+    message_type = header[19:20]
+    return conversation_id, message_id, message_type
+
+
+def serialize_data(data: object) -> bytes:
+    """Turn `data` into a bytes object.
+
+    Due to json serialization, data must not contain a bytes object!
+    """
+    if isinstance(data, BaseModel):
+        return data.json().encode()
```

Comment:
not sure what this here is doing tbh. how do we end up with a `BaseModel` being fed into this? More context would be great!

Reply:
The json-rpc-objects are subclasses of BaseModel.
If we have such an object (a Request, a Reply, an Error...), we want to use their json method. Otherwise, we use the json module.

Reply:
I import the appropriate objects and verify them, that is more explicit.